src/bin/extensions.ml

```
@@ -0,0 +1,125 @@
+  let all_extensions =
+    Hashtbl.fold
+      (fun extension_name () exts ->
+        extension_name :: exts)
+      builtin_extensions
+      []
+  in
+  let all_extensions =
+    List.fold_left (fun all_extensions plugin ->
+      if Hashtbl.mem builtin_extensions plugin
+      then
+        all_extensions
+      else
+        plugin :: all_extensions
+    ) all_extensions (Dolmen.Sites.Plugins.Plugins.list ())
+  in
+  List.fast_sort
+    (fun e1 e2 -> String.compare (name e1) (name e2))
+    all_extensions
+
+let loaded_plugins =
+  Hashtbl.create 17
+
+let dynlink_error =
+  let open Dolmen_loop in
+  Report.Warning.mk ~mnemonic:"dynlink-plugin-error"
+    ~message:(fun ppf (plugin, err) ->
+      Fmt.pf ppf "Unable to load plugin library '%s':@ %s"
+        plugin (Dynlink.error_message err))
+    ~name:"Dynlink error" ()
+
+let list_available_extensions (_plugin, _exn) =
+  Some (
+    Format.dprintf
+      "Available extensions are:@ %a"
+      Fmt.(list ~sep:Fmt.comma (box pp)) (list ())
+  )
+
+let dune_plugin_error =
+  Dolmen_loop.Report.Error.mk ~mnemonic:"dune-plugin-error"
+    ~hints:[list_available_extensions]
+    ~message:(fun ppf (plugin, exn) ->
+      Fmt.pf ppf "Unknown error while loading plugin '%s':@ %s"
+        plugin (Printexc.to_string exn))
+    ~name:"Dune plugin error" ()
+
+let load_plugin_or_fail plugin st =
+  if Hashtbl.mem loaded_plugins plugin then st
+  else
+    (* Ideally, we'd like to check for a conflict with a dune plugin here,
+       but there is no nice way to check if a plugin of a given name exists
+       without actually trying to load the plugin. *)
+    if Hashtbl.mem builtin_extensions plugin
+    then (
+      Hashtbl.replace loaded_plugins plugin ();
+      st
+    )
+    else
+      match Dolmen.Sites.Plugins.Plugins.load plugin with
+      | () ->
+        Hashtbl.replace loaded_plugins plugin ();
+        st
+      | exception Dynlink.Error err ->
+        Loop.State.warn st dynlink_error (plugin, err)
```

In that case, are we sure that the plugin has been actually loaded ? Else this could probably be an error ?

---

I think we are sure that the plugin has *not* been loaded. It's probably better to always fail if the plugin is not loaded though, rather than let execution continue when a plugin that was explicitly requested by the user failed to load. Will change to an error.